Fix The Princess Takes Flight (Saga first chapter zcc)

ssk97 opened the pull request:
Currently, The Princess Takes Flight will not return the creature it exiled. This is because for all sagas, the first chapter ability triggers occurs while the saga is entering the battlefield, before its zcc is updated.

This is an attempt at fixing this by incrementing the zcc of any triggered ability that triggers while a permanent is entering, need to investigate further to see if there's a better solution. Also need to investigate if this solution works when the saga is flickered and other unusual scenarios.

xenohedron commented:
Huh, no test failures. I wouldn't be surprised if some custom effects manually offset already to compensate, which might get broken by this change (I vaguely remember doing something like that before). Better to fix in the general case as long as we understand the ramifications.

JayDi85 commented:
There are few places with same zcc init code logic, search by `getSourceObjectZoneChangeCounter() == 0`. Must be research. I think it must use same logic in all places, not for some triggers only.

Few examples:

1. Mana triggered abilities:
![shot_240108_142043](https://github.com/magefree/mage/assets/8344157/38538eeb-d198-4019-8f6d-c025c2ff8818)

2. Normal abilities:
![shot_240108_141443](https://github.com/magefree/mage/assets/8344157/da337c15-28fe-4dee-9d36-c953ba6fca24)

3. Some counters related code:
![shot_240108_142105](https://github.com/magefree/mage/assets/8344157/f6c8b9ca-5785-4d86-b5a3-b66d9bf2b4b4)

4. Some cards mor related code:
![shot_240108_142319](https://github.com/magefree/mage/assets/8344157/05f0e212-29b8-4647-ba19-00652198750e)

JayDi85 commented:
Possible solution: add helper method like "initZCC" and sync it with real/entering object on zcc == 0 (on activate, on added, in other places).

ssk97 commented:
> Possible solution: add helper method like "initZCC" and sync it with real/entering object on zcc == 0 (on activate, on added, in other places).

Ok, I've unified the zcc updating logic, I think that's what you wanted? I'm not sure if I should've included the "always update zcc" instances but I figured it didn't hurt and the tests passed. I you want me to revert those just let me know.

JayDi85 commented on Mage/src/main/java/mage/abilities/AbilityImpl.java:
```diff
@@ -1328,6 +1326,26 @@ public void setSourceObjectZoneChangeCounter(int sourceObjectZoneChangeCounter)
+    @Override
+    public void updateSourceObjectZoneChangeCounter(Game game, boolean force) {
+        if (!(this instanceof MageSingleton) && (force || sourceObjectZoneChangeCounter == 0 )) {
+            setSourceObjectZoneChangeCounter(getCurrentSourceZCC(game));
```
That's force param smells bad. Need research on usage and docs for it. Maybe it can be renamed to "copiedAbility" or like that (or maybe not). Maybe it can be removed at all (or maybe not).

Also `updateSourceObjectZoneChangeCounter` must be renamed to `initSourceObjectZoneChangeCounter`

ssk97 replied:
It definitely can't be entirely removed without additional work, I tried that just to see and there were multiple test failures. I'll investigate further.

JayDi85 commented on Mage/src/main/java/mage/game/GameImpl.java:
```diff
@@ -2081,7 +2077,7 @@ public UUID addDelayedTriggeredAbility(DelayedTriggeredAbility delayedAbility, A
         if (source != null) {
-            newAbility.setSourceObjectZoneChangeCounter(getState().getZoneChangeCounter(source.getSourceId()));
+            newAbility.updateSourceObjectZoneChangeCounter(this, true);
             newAbility.setSourcePermanentTransformCount(this);
```
Looks like `setSourcePermanentTransformCount` miss from all other `setSourceObjectZoneChangeCounter` usages. It can be a bug -- need research. `setSourcePermanentTransformCount` introduced with day/bound so dev might have forgotten to add it in other places.

JayDi85 commented on Mage/src/main/java/mage/abilities/AbilityImpl.java:
```diff
@@ -1328,6 +1326,26 @@ public void setSourceObjectZoneChangeCounter(int sourceObjectZoneChangeCounter)
+        }
+    }
+
+    private int getCurrentSourceZCC(Game game){
```
Must be renamed to `getCurrentSourceObjectZoneChangeCounter`

ssk97 commented:
Sorry for the delay, I've made the requested name changes and added an additional test for Sigarda's Splendor to confirm its behavior.

I think it would be more "correct" to make abilities like Sigarda's Splendor use the permanent ZCC instead of the spell's, but there's just too many card implementations that would have to be changed for that to be reasonable so there's no change there.

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/mid/SigardasSplendorTest.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @author notgreat
+ */
+public class SigardasSplendorTest extends CardTestPlayerBase {
```
looks like this file is indented with tabs? spaces are standard code style here

ssk97 replied:
Oops, I copied from the other `MID` test which also happened to be using tabs. Will fix both of them.

xenohedron reviewed (commented):
can you please add some documentation on the usage of that "force" parameter?

xenohedron commented:
Well, I still don't understand when code should or shouldn't use the force parameter.

But this fixes a bug that affects several existing cards and a few pending implementations, so I would like to move forward with this.

@JayDi85 you marked as don't merge - do you still have objections?

ssk97 commented:
> Well, I still don't understand when code should or shouldn't use the force parameter.

I spent a while trying to figure out the details, I've got some info but not a full picture. That's why I'm not changing the logic there at all, just copying/centralizing the already-existing code.

JayDi85 commented:
> you marked as don't merge - do you still have objections?

It’s still unfinished and can’t be merge.

Example of author’s comment about force param:

> It definitely can't be entirely removed without additional work, I tried that just to see and there were multiple test failures. I'll investigate further.

Another unresolved example:

Ignored bug with missing sync code of transformCount

And another:

Global zcc logic was changed for all effects without real effects research and fix cause “it would be annoying to update all of them”. Now it works in one use cases and doesn’t work in another use cases (and you don’t know how to use it now). Also ignored/un-researched code that uses zcc == 0.

P.S.
After today review found another one un-researched problem with `getSourceObjectIfItStillExists` - it also changed and now return entering permanent instead real. So all usages must checked and fixed too (if require). And new logic will work with getSourceObjectIfItStillExists usage only, but not other zcc related code. So some effects will see a permanent, another one will not see (hello blood moon #4202).

P.S. Sometime “simple fix” requires much more fixes and researches. Extreme example: miss morph image bug from #11622 transforms to face down rework from #11873.

ssk97 commented:
> Example of author’s comment about force param:
> > It definitely can't be entirely removed without additional work, I tried that just to see and there were multiple test failures. I'll investigate further.
> 
> Another unresolved example:
> 
> Ignored bug with missing sync code of transformCount

Those are existing issues with the code that, if they have bugs, are completely unaffected by my changes. I just moved the code into a single function. If you'd like me to add the `transformCount` updating I will, I just didn't want to mess with that without fully understanding it. Testing indicated that updating the count at all times would not cause problems, but I wanted to do the minimum possible to fix the bugs.

> Global zcc logic was changed for all effects without real effects research and fix cause “it would be annoying to update all of them”.

The only abilities affected by this change are those which trigger as a result of effects that happen as a permanent is entering the battlefield. In particular, abilities that trigger as a result of adding a counter for permanents that enter with a counter on them.

I think that it would be better if everything entering the battlefield referred to the zcc of the permanent instead of the spell, but since a significant number of cards were written under the current situation it's not worth updating as it likely wouldn't fix any bugs. Those _always_ refer to the spell zcc whereas the ones affected by this change currently trigger usually with the permanent's zcc but sometimes the spell's in situations like Chapter 1 of a saga.

> After today review found another one un-researched problem with getSourceObjectIfItStillExists - it also changed and now return entering permanent instead real. 

The current situation is that relevant abilities do not see any permanents, since the ZCC refers to the spell on the stack. This changes it so that in the moderately common situation of sagas entering the battlefield, the ZCC is now referring to the entering permanent (which at resolution of the ability will be a fully resolved permanent) rather than that of the spell. Note that `getSourceObjectIfItStillExists` is not just for permanents on the battlefield, it is used for everything; spells on the stack, permanents still entering the battlefield, etc. So long as the zcc matches, it works.

Susucre commented:
I also agree that the saga's effect should use the zcc of the permanent.
I'll make a list of the sagas that would require checking after such a change, as they may have workaround the issue. And start create basic tests for them to ease the change.

Grath commented:
> It’s better to improve and fix specific saga instead inject new zcc hack into existing bunch of hacks. It’s a type of work that will break cards in any use cases, cause some cards created for one zcc logic and another cards for fully different logic.

This needs to be fixed properly, and doing zcc hacks on saga first abilities is the broken and incorrect way to do it. First abilities can trigger 'naturally' as the saga is entering **and** they can trigger while the saga is in play by simply removing the Lore counters from the saga, so the ability needs to work correctly whether or not the saga is already in play. The only correct solutions I can think of are:

- Have zcc offset hacks in every saga first ability (where the zcc is relevant) that activate if and only if the saga is currently entering the battlefield and otherwise don't offset the zcc.
- This approach where the ability's zone change counter is globally adjusted to relate to the permanent as it is when it's in play.

Fixing this at an engine level reduces the number of zcc offset hacks overall and makes it easier to maintain and implement these types of abilities.

JayDi85 commented:
Current PR is unfinished and there are many miss places/fixes, see my and xeno comments above. I recommend to review/rewrite instead merge two years staled/unfinished PR.

ssk97 commented:
> Current PR is unfinished and there are many miss places/fixes, see my and xeno comments above. I recommend to review/rewrite it again instead merge two years staled/unfinished PR.

AFAIK the only things that were left unresolved was that I didn't do extra research to figure out if there was a better solution to zcc updating than the "force" parameter, and trying to understand if `setSourcePermanentTransformCount` needed to be used in more places. In both instances, my code had no change from the current behavior, so while it could potentially fix more bugs, it wouldn't add any new ones.

There was one relevant engine change since I wrote the code, d8959f1588f0556ba4f4604d5c0dbf98350333b8 added special zcc handling for `DelayedTriggeredAbility`, which required some minor changes that I have now done. This does also make it a little uglier, as the zcc updating is no longer completely unified.

I've removed the offsets for the sagas @Susucre added tests for, I have not checked the other remaining sagas in that list.

Susucre commented:
@ssk97
I'm done with doing basic tests on the zcc-related sagas.
Got more ignored tests to enable here for Vault 13, Summon: Ixion and The War Games.

ssk97 commented:
Ok, Vault 13 had some pretty major problems. I've fixed them here and also did a minor rework of some related cards, if we decide to throw away this PR I'll push that rework into master separately.

xenohedron commented on Mage/src/main/java/mage/target/targetadjustment/ForEachOpponentTargetsAdjuster.java:
```diff
@@ -11,12 +11,14 @@
  */
 public class ForEachOpponentTargetsAdjuster extends GenericTargetAdjuster {
     private final boolean owner;
+    private final boolean includeSelf; //Makes this a "For Each Player" adjuster
```
I don't like this sort of backwards naming convention. I could understand onlyOpponents as a subset of all players, but it is not intuitive to use a class with "opponent" in the name for all players. Personally I would prefer two separately named classes which keeps the constructors simple.

ssk97 replied:
I prefer keeping things as centralized as possible, especially when there's only a couple dozen cards using it. I do agree that `ForEachPlayer` and an `onlyOpponents` argument are better, will make that change.

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/PutOnLibraryTargetEffect.java:
```diff
@@ -63,6 +63,7 @@ public boolean apply(Game game, Ability source) {
                         break;
                     case GRAVEYARD:
+                    case EXILED:
                         Card graveyardCard = game.getCard(targetId);
```
worth renaming the variable

also below, unrelated to this PR, null check on getSpell?

xenohedron commented:
I recommend separate fresh commits to master for the unrelated fixes (or can be PRs if any other feedback desired):
* new tests (also, don't have two separate classes for The Princess Takes Flight tests)
* PutOnLibraryTargetEffect support exile zone
* ForEachPlayerTargetAdjuster
* any other card-specific fixes not fundamentally related to zcc offset

And then after that, fresh PR from master for the zcc change, with only the code necessary to pass the ignored tests
